File: src-tauri/src/parser/parsers/cities.rs
// Cities parser - pure parsing without database dependency

use crate::parser::game_data::CityData;
use crate::parser::xml_loader::{XmlDocument, XmlNodeExt};
use crate::parser::Result;

/// Parse cities to structs (no DB dependency)
pub fn parse_cities_struct(doc: &XmlDocument) -> Result<Vec<CityData>> {
    let root = doc.root_element();
    let mut cities = Vec::new();

    // Find all City elements as direct children of Root
    for city_node in root.children().filter(|n| n.has_tag_name("City")) {
        let xml_id = city_node.req_attr("ID")?.parse::<i32>()?;

        // Player ID - filter out -1 for cities in anarchy/being captured
        let player_xml_id = city_node.req_attr("Player")?.parse::<i32>()?;
        let player_xml_id = if player_xml_id >= 0 {
            Some(player_xml_id)
        } else {
            None
        };

        let tile_xml_id = city_node.req_attr("TileID")?.parse::<i32>()?;

        // City name: older saves use "NameType", newer saves use "Name"
        let city_name = city_node
            .opt_child_text("NameType")
            .or_else(|| city_node.opt_child_text("Name"))
            .unwrap_or("Unknown City")
            .to_string();

        let founded_turn = city_node.req_attr("Founded")?.parse::<i32>()?;

        // Optional fields
        let family = city_node.opt_attr("Family").map(|s| s.to_string());

        // Capital status is indicated by presence of <Capital /> element
        let is_capital = city_node.children().any(|n| n.has_tag_name("Capital"));

        // Population
        let citizens = city_node
            .opt_child_text("Citizens")
            .and_then(|s| s.parse::<i32>().ok())
            .unwrap_or(1);

        // Leadership
        let governor_xml_id = city_node
            .opt_child_text("GovernorID")
            .and_then(|s| s.parse::<i32>().ok());

        // Production
        let hurry_civics_count = city_node
            .opt_child_text("HurryCivicsCount")
            .and_then(|s| s.parse::<i32>().ok())
            .unwrap_or(0);

        let hurry_money_count = city_node
            .opt_child_text("HurryMoneyCount")
            .and_then(|s| s.parse::<i32>().ok())
            .unwrap_or(0);

        let specialist_count = city_node
            .opt_child_text("SpecialistProducedCount")
            .and_then(|s| s.parse::<i32>().ok())
            .unwrap_or(0);

        // First owner tracking
        let first_owner_player_xml_id = city_node
            .opt_child_text("FirstPlayer")
            .and_then(|s| s.parse::<i32>().ok());

        // Last owner tracking
        let last_owner_player_xml_id = city_node
            .opt_child_text("LastPlayer")
            .and_then(|s| s.parse::<i32>().ok());

        // Governor turn
        let governor_turn = city_node
            .opt_child_text("GovernorTurn")
            .and_then(|s| s.parse::<i32>().ok());

        // Additional hurry metrics
        let hurry_training_count = city_node
            .opt_child_text("HurryTrainingCount")
            .and_then(|s| s.parse::<i32>().ok())
            .unwrap_or(0);

        let hurry_population_count = city_node
            .opt_child_text("HurryPopulationCount")
            .and_then(|s| s.parse::<i32>().ok())
            .unwrap_or(0);

        // Growth and production counts
        let growth_count = city_node
            .opt_child_text("GrowthCount")
            .and_then(|s| s.parse::<i32>().ok())
            .unwrap_or(0);

        let unit_production_count = city_node
            .opt_child_text("UnitProductionCount")
            .and_then(|s| s.parse::<i32>().ok())
            .unwrap_or(0);

        let buy_tile_count = city_node
            .opt_child_text("BuyTileCount")
            .and_then(|s| s.parse::<i32>().ok())
            .unwrap_or(0);

        cities.push(CityData {
            xml_id,
            city_name,
            founded_turn,
            player_xml_id,
            tile_xml_id,
            family,
            first_owner_player_xml_id,
            last_owner_player_xml_id,
            is_capital,
            citizens,
            governor_xml_id,
            governor_turn,
            hurry_civics_count,
            hurry_money_count,
            hurry_training_count,
            hurry_population_count,
            specialist_count,
            growth_count,
            unit_production_count,
            buy_tile_count,
        });
    }

    Ok(cities)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::xml_loader::parse_xml;

    #[test]
    fn test_parse_cities_struct_basic() {
        let xml = r#"<Root GameId="test-123">
            <City ID="0" Player="0" TileID="100" Founded="1">
                <Name>Test City</Name>
                <Citizens>3</Citizens>
                <Capital />
            </City>
        </Root>"#;

        let doc = parse_xml(xml.to_string()).unwrap();
        let cities = parse_cities_struct(&doc).unwrap();

        assert_eq!(cities.len(), 1);
        assert_eq!(cities[0].xml_id, 0);
        assert_eq!(cities[0].city_name, "Test City");
        assert_eq!(cities[0].player_xml_id, Some(0));
        assert_eq!(cities[0].tile_xml_id, 100);
        assert_eq!(cities[0].founded_turn, 1);
        assert_eq!(cities[0].is_capital, true);
        assert_eq!(cities[0].citizens, 3);
    }

    #[test]
    fn test_parse_cities_struct_anarchy() {
        let xml = r#"<Root GameId="test-123">
            <City ID="0" Player="-1" TileID="100" Founded="1">
                <Name>Anarchy City</Name>
            </City>
        </Root>"#;

        let doc = parse_xml(xml.to_string()).unwrap();
        let cities = parse_cities_struct(&doc).unwrap();

        assert_eq!(cities.len(), 1);
        assert_eq!(cities[0].player_xml_id, None); // -1 filtered out
    }

    #[test]
    fn test_parse_cities_struct_multiple() {
        let xml = r#"<Root GameId="test-123">
            <City ID="0" Player="0" TileID="100" Founded="1">
                <Name>City One</Name>
            </City>
            <City ID="1" Player="1" TileID="200" Founded="5">
                <Name>City Two</Name>
                <Citizens>5</Citizens>
            </City>
        </Root>"#;

        let doc = parse_xml(xml.to_string()).unwrap();
        let cities = parse_cities_struct(&doc).unwrap();

        assert_eq!(cities.len(), 2);
        assert_eq!(cities[0].city_name, "City One");
        assert_eq!(cities[1].city_name, "City Two");
        assert_eq!(cities[1].citizens, 5);
    }

    #[test]
    fn test_parse_cities_first_player() {
        let xml = r#"<Root GameId="test">
            <City ID="0" Player="0" TileID="100" Founded="1">
                <Name>Test</Name>
                <FirstPlayer>2</FirstPlayer>
            </City>
        </Root>"#;
        let doc = parse_xml(xml.to_string()).unwrap();
        let cities = parse_cities_struct(&doc).unwrap();
        assert_eq!(cities[0].first_owner_player_xml_id, Some(2));
    }

    #[test]
    fn test_parse_cities_specialist_produced_count() {
        let xml = r#"<Root GameId="test">
            <City ID="0" Player="0" TileID="100" Founded="1">
                <Name>Test</Name>
                <SpecialistProducedCount>24</SpecialistProducedCount>
            </City>
        </Root>"#;
        let doc = parse_xml(xml.to_string()).unwrap();
        let cities = parse_cities_struct(&doc).unwrap();
        assert_eq!(cities[0].specialist_count, 24);
    }

    #[test]
    fn test_parse_cities_all_hurry_metrics() {
        let xml = r#"<Root GameId="test">
            <City ID="0" Player="0" TileID="100" Founded="1">
                <Name>Test</Name>
                <HurryCivicsCount>3</HurryCivicsCount>
                <HurryTrainingCount>2</HurryTrainingCount>
                <HurryMoneyCount>5</HurryMoneyCount>
                <HurryPopulationCount>4</HurryPopulationCount>
            </City>
        </Root>"#;
        let doc = parse_xml(xml.to_string()).unwrap();
        let cities = parse_cities_struct(&doc).unwrap();
        assert_eq!(cities[0].hurry_civics_count, 3);
        assert_eq!(cities[0].hurry_training_count, 2);
        assert_eq!(cities[0].hurry_money_count, 5);
        assert_eq!(cities[0].hurry_population_count, 4);
    }

    #[test]
    fn test_parse_cities_new_fields() {
        let xml = r#"<Root GameId="test">
            <City ID="0" Player="0" TileID="100" Founded="1">
                <Name>Test</Name>
                <GovernorTurn>45</GovernorTurn>
                <GrowthCount>12</GrowthCount>
                <UnitProductionCount>8</UnitProductionCount>
                <LastPlayer>0</LastPlayer>
                <BuyTileCount>3</BuyTileCount>
            </City>
        </Root>"#;
        let doc = parse_xml(xml.to_string()).unwrap();
        let cities = parse_cities_struct(&doc).unwrap();
        assert_eq!(cities[0].governor_turn, Some(45));
        assert_eq!(cities[0].growth_count, 12);
        assert_eq!(cities[0].unit_production_count, 8);
        assert_eq!(cities[0].last_owner_player_xml_id, Some(0));
        assert_eq!(cities[0].buy_tile_count, 3);
    }
}

File: src/lib/parser/parsers/cities.ts
// City entity + 8 sub-entity parsers.
//
// Per spec §394 these live in one consolidated module rather than one per
// collection — all 9 share the same per-City iteration shape and benefit
// from being read together.

import {
	asArray,
	getElementChildren,
	isElement,
	optAttrStr,
	optInt,
	optStr,
	parseNameKeyedIntMap,
	parsePrefixedKeyedIntMap,
	requireInt,
	requireStr,
} from "../parse-xml.js";

// ---------- Types ----------

export interface City {
	xmlId: number;
	cityName: string;
	foundedTurn: number;
	playerXmlId: number | null;
	tileXmlId: number;
	family: string | null;
	// Per-player family associations from <PlayerFamily> (<P.X>FAMILY_Y</P.X>).
	// Retains each owning player's family across a capture — a city's family
	// only changes when another nation conquers it, so this is the historical
	// family record for the map's turn slider. Same shape as unit families
	// (parsers/units.ts parseUnitFamilies).
	playerFamilies: { playerXmlId: number; familyName: string }[];
	firstOwnerPlayerXmlId: number | null;
	lastOwnerPlayerXmlId: number | null;
	isCapital: boolean;
	citizens: number;
	governorXmlId: number | null;
	governorTurn: number | null;
	hurryCivicsCount: number;
	hurryMoneyCount: number;
	hurryTrainingCount: number;
	hurryPopulationCount: number;
	specialistCount: number;
	growthCount: number;
	unitProductionCount: number;
	buyTileCount: number;
	// Per-city project completions from <ProjectCount> (PROJECT_X → count) —
	// city projects like Archives and Forums that carry city effects.
	// (Religion presence is parsed separately — parseCityReligions.)
	projectCounts: { project: string; count: number }[];
}

export interface CityProductionItem {
	cityXmlId: number;
	queuePosition: number;
	buildType: string;
	itemType: string;
	progress: number;
	isRepeat: boolean;
}

export interface CityProjectCompleted {
	cityXmlId: number;
	projectType: string;
	count: number;
}

export interface CityProjectCount {
	cityXmlId: number;
	projectType: string;
	count: number;
}

export interface CityEnemyAgent {
	cityXmlId: number;
	enemyPlayerXmlId: number;
	agentCharacterXmlId: number | null;
	placedTurn: number | null;
	agentTileXmlId: number | null;
}

export interface CityLuxury {
	cityXmlId: number;
	resource: string;
	importedTurn: number;
}

export interface CityYield {
	cityXmlId: number;
	yieldType: string;
	progress: number;
}

export interface CityReligion {
	cityXmlId: number;
	religion: string;
}

export interface CityCulture {
	cityXmlId: number;
	teamId: number;
	cultureLevel: string | null;
	happinessLevel: number;
}

// ---------- Helpers ----------

function eachCity(
	root: Record<string, unknown>,
): Generator<[number, Record<string, unknown>]> {
	return (function* () {
		for (const node of asArray(root.City) as unknown[]) {
			if (!isElement(node)) continue;
			const cityXmlId = requireInt(node["@_ID"], "City.ID");
			yield [cityXmlId, node];
		}
	})();
}

// ---------- Cities core ----------

export function parseCities(root: Record<string, unknown>): City[] {
	const cities: City[] = [];

	for (const [, node] of eachCity(root)) {
		const xmlId = requireInt(node["@_ID"], "City.ID");
		const playerRaw = requireInt(node["@_Player"], "City.Player");
		const playerXmlId = playerRaw >= 0 ? playerRaw : null;
		const tileXmlId = requireInt(node["@_TileID"], "City.TileID");
		const foundedTurn = requireInt(node["@_Founded"], "City.Founded");

		// Older saves use <NameType>, newer use <Name>; default to literal
		// "Unknown City" if neither resolves to a non-empty string.
		const cityName =
			optStr(node.NameType) ?? optStr(node.Name) ?? "Unknown City";

		// is_capital: presence-of-element check. fast-xml-parser represents
		// <Capital/> as `Capital: ""`; the key is present iff the element is.
		const isCapital = "Capital" in node;

		// unit_production_count: prefer aggregate <UnitProductionCount>,
		// fall back to sum of <UnitProductionCounts> children, default 0.
		const aggregateUpc = optInt(node.UnitProductionCount);
		const unitProductionCount =
			aggregateUpc !== null
				? aggregateUpc
				: sumNamedIntChildren(node.UnitProductionCounts);

		// <PlayerFamily> holds <P.X>FAMILY_Y</P.X> for every player who has
		// owned this city — the same string-valued prefix-keyed shape as unit
		// families. Skip non-P.* tags, unparseable player
		// IDs, and empty values.
		const playerFamilies: { playerXmlId: number; familyName: string }[] = [];
		const familyNode = node.PlayerFamily;
		if (isElement(familyNode)) {
			for (const [tag, value] of getElementChildren(familyNode)) {
				if (!tag.startsWith("P.")) continue;
				const pid = parseInt(tag.slice(2), 10);
				if (Number.isNaN(pid)) continue;
				if (typeof value !== "string" || value === "") continue;
				playerFamilies.push({ playerXmlId: pid, familyName: value });
			}
		}

		const projectCounts = [...parseNameKeyedIntMap(node.ProjectCount)].map(
			([project, count]) => ({ project, count }),
		);

		cities.push({
			xmlId,
			cityName,
			foundedTurn,
			playerXmlId,
			tileXmlId,
			family: optAttrStr(node["@_Family"]),
			playerFamilies,
			firstOwnerPlayerXmlId: optInt(node.FirstPlayer),
			lastOwnerPlayerXmlId: optInt(node.LastPlayer),
			isCapital,
			citizens: optInt(node.Citizens) ?? 1,
			governorXmlId: optInt(node.GovernorID),
			governorTurn: optInt(node.GovernorTurn),
			hurryCivicsCount: optInt(node.HurryCivicsCount) ?? 0,
			hurryMoneyCount: optInt(node.HurryMoneyCount) ?? 0,
			hurryTrainingCount: optInt(node.HurryTrainingCount) ?? 0,
			hurryPopulationCount: optInt(node.HurryPopulationCount) ?? 0,
			specialistCount: optInt(node.SpecialistProducedCount) ?? 0,
			growthCount: optInt(node.GrowthCount) ?? 0,
			unitProductionCount,
			buyTileCount: optInt(node.BuyTileCount) ?? 0,
			projectCounts,
		});
	}

	return cities;
}

function sumNamedIntChildren(node: unknown): number {
	if (!isElement(node)) return 0;
	let sum = 0;
	for (const value of parseNameKeyedIntMap(node).values()) sum += value;
	return sum;
}

// ---------- Production queue ----------

export function parseCityProductionQueue(
	root: Record<string, unknown>,
): CityProductionItem[] {
	const items: CityProductionItem[] = [];

	for (const [cityXmlId, node] of eachCity(root)) {
		const queueNode = node.BuildQueue;
		if (!isElement(queueNode)) continue;

		let queuePosition = 0;
		for (const qi of asArray(queueNode.QueueInfo) as unknown[]) {
			if (!isElement(qi)) continue;
			items.push({
				cityXmlId,
				queuePosition: queuePosition++,
				buildType: requireStr(qi.Build, "QueueInfo.Build"),
				itemType: requireStr(qi.Type, "QueueInfo.Type"),
				progress: optInt(qi.Progress) ?? 0,
				// IsRepeat parses as int, then non-zero. Default false.
				isRepeat: (optInt(qi.IsRepeat) ?? 0) !== 0,
			});
		}
	}

	return items;
}

// ---------- Projects completed ----------

export function parseCityProjectsCompleted(
	root: Record<string, unknown>,
): CityProjectCompleted[] {
	const out: CityProjectCompleted[] = [];

	for (const [cityXmlId, node] of eachCity(root)) {
		const completedNode = node.CompletedBuild;
		if (!isElement(completedNode)) continue;

		const counts = new Map<string, number>();
		for (const qi of asArray(completedNode.QueueInfo) as unknown[]) {
			if (!isElement(qi)) continue;
			// Missing Build/Type both default to literal "UNKNOWN". A typical
			// CompletedBuild record has both, but be defensive — that exact
			// spelling is what Rust emits.
			const buildType = optStr(qi.Build) ?? "UNKNOWN";
			const itemType = optStr(qi.Type) ?? "UNKNOWN";
			const projectType = `${buildType}.${itemType}`;
			counts.set(projectType, (counts.get(projectType) ?? 0) + 1);
		}

		for (const [projectType, count] of counts) {
			out.push({ cityXmlId, projectType, count });
		}
	}

	return out;
}

// ---------- Project counts ----------

export function parseCityProjectCounts(
	root: Record<string, unknown>,
): CityProjectCount[] {
	const out: CityProjectCount[] = [];

	for (const [cityXmlId, node] of eachCity(root)) {
		const projectCountNode = node.ProjectCount;
		if (!isElement(projectCountNode)) continue;

		// Rust filters count > 0, so unparseable / zero-count entries are
		// dropped. parseNameKeyedIntMap already drops unparseable.
		for (const [projectType, count] of parseNameKeyedIntMap(projectCountNode)) {
			if (count > 0) {
				out.push({ cityXmlId, projectType, count });
			}
		}
	}

	return out;
}

// ---------- Enemy agents ----------

export function parseCityEnemyAgents(
	root: Record<string, unknown>,
): CityEnemyAgent[] {
	const out: CityEnemyAgent[] = [];

	for (const [cityXmlId, node] of eachCity(root)) {
		const turns = parsePrefixedKeyedIntMap(node.AgentTurn, "P.");
		const chars = parsePrefixedKeyedIntMap(node.AgentCharacterID, "P.");
		const tiles = parsePrefixedKeyedIntMap(node.AgentTileID, "P.");

		const enemies = new Set<number>([
			...turns.keys(),
			...chars.keys(),
			...tiles.keys(),
		]);

		for (const enemyPlayerXmlId of enemies) {
			out.push({
				cityXmlId,
				enemyPlayerXmlId,
				agentCharacterXmlId: chars.get(enemyPlayerXmlId) ?? null,
				placedTurn: turns.get(enemyPlayerXmlId) ?? null,
				agentTileXmlId: tiles.get(enemyPlayerXmlId) ?? null,
			});
		}
	}

	return out;
}

// ---------- Luxuries ----------

export function parseCityLuxuries(root: Record<string, unknown>): CityLuxury[] {
	const out: CityLuxury[] = [];

	for (const [cityXmlId, node] of eachCity(root)) {
		const luxuryNode = node.LuxuryTurn;
		if (!isElement(luxuryNode)) continue;

		for (const [resource, importedTurn] of parseNameKeyedIntMap(luxuryNode)) {
			out.push({ cityXmlId, resource, importedTurn });
		}
	}

	return out;
}

// ---------- Yields ----------

export function parseCityYields(root: Record<string, unknown>): CityYield[] {
	const out: CityYield[] = [];

	for (const [cityXmlId, node] of eachCity(root)) {
		const yieldNode = node.YieldProgress;
		if (!isElement(yieldNode)) continue;

		for (const [yieldType, progress] of parseNameKeyedIntMap(yieldNode)) {
			out.push({ cityXmlId, yieldType, progress });
		}
	}

	return out;
}

// ---------- Religions ----------

export function parseCityReligions(
	root: Record<string, unknown>,
): CityReligion[] {
	const out: CityReligion[] = [];

	for (const [cityXmlId, node] of eachCity(root)) {
		const religionNode = node.Religion;
		if (!isElement(religionNode)) continue;

		// Just collect element child names — values aren't used here.
		for (const [religion] of getElementChildren(religionNode)) {
			out.push({ cityXmlId, religion });
		}
	}

	return out;
}

// ---------- Culture ----------

export function parseCityCulture(root: Record<string, unknown>): CityCulture[] {
	const out: CityCulture[] = [];

	for (const [cityXmlId, node] of eachCity(root)) {
		// TeamCulture: <T.X>CULTURE_LEGENDARY</T.X> string-valued. Single-use
		// shape (string-valued prefix-keyed map) — inline the walk rather
		// than adding a one-off helper.
		const teamCulture = new Map<number, string>();
		const cultureNode = node.TeamCulture;
		if (isElement(cultureNode)) {
			for (const [tagName, value] of getElementChildren(cultureNode)) {
				if (!tagName.startsWith("T.")) continue;
				if (typeof value !== "string") continue;
				const teamId = parseInt(tagName.slice(2), 10);
				if (Number.isNaN(teamId)) continue;
				teamCulture.set(teamId, value);
			}
		}

		// Happiness: prefer <TeamHappinessLevel> (newer), fall back to
		// <TeamDiscontentLevel> (legacy 2022 saves).
		const happinessRaw =
			node.TeamHappinessLevel ?? node.TeamDiscontentLevel ?? null;
		const teamHappiness = parsePrefixedKeyedIntMap(happinessRaw, "T.");

		const teams = new Set<number>([
			...teamCulture.keys(),
			...teamHappiness.keys(),
		]);

		for (const teamId of teams) {
			out.push({
				cityXmlId,
				teamId,
				cultureLevel: teamCulture.get(teamId) ?? null,
				happinessLevel: teamHappiness.get(teamId) ?? 0,
			});
		}
	}

	return out;
}
